File: xkcd/src/main/java/xyz/jienan/xkcd/XkcdExplainUtil.java
package xyz.jienan.xkcd;

import android.text.TextUtils;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import okhttp3.ResponseBody;

import static xyz.jienan.xkcd.Const.URI_XKCD_EXPLAIN_EDIT;

public class XkcdExplainUtil {
    public static String getExplainFromHtml(ResponseBody responseBody, String url) throws IOException {
        Document doc = Jsoup.parse(responseBody.string());
        Elements newsHeadlines = doc.select("h2");
        for (Element headline : newsHeadlines) {
            if (isH2ByType(headline, "Explanation")) {
                Element element = headline.nextElementSibling();
                StringBuilder htmlResult = new StringBuilder();
                while (!"h2".equals(element.nodeName())) {
                    if ("h3".equals(element.nodeName()) || "h4".equals(element.nodeName())) {
                        Elements elements = element.getElementsByClass("editsection");
                        if (elements != null && elements.size() > 0) {
                            elements.remove();
                        }
                    }
                    if (element.tagName().equals("p"))
                        if (element.toString().contains("<i>citation needed</i>")) {
                            List<Node> nodes = new ArrayList<>();
                            for (Node node : element.childNodes()) {
                                if ("sup".equals(node.nodeName()) && node.toString().contains("<i>citation needed</i>")) {
                                    nodes.add(node);
                                }
                            }
                            for (Node node : nodes) {
                                node.remove();
                            }
                        }
                    for (Element child : element.getAllElements()) {
                        if ("a".equals(child.tagName()) && child.hasAttr("href")) {
                            String href = child.attr("href");
                            if (!TextUtils.isEmpty(href)) {
                                if (href.startsWith("/wiki")) {
                                    child.attr("href", "https://www.explainxkcd.com" + href);
                                } else if (href.startsWith("#")) {
                                    child.attr("href", url + href);
                                } else if (href.startsWith("//www.explainxkcd") && href.endsWith("action=edit")) {
                                    child.attr("href", URI_XKCD_EXPLAIN_EDIT);
                                }
                            }
                        }
                        if ("tbody".equals(child.tagName())) {
                            for (Element tableElement : child.children()) {
                                tableElement.append("<br />");
                            }
                        }
                    }
                    htmlResult.append(element.toString());
                    Node node = element.nextSibling();
                    while (!(node instanceof Element)) {
                        htmlResult.append(node.toString());
                        node = node.nextSibling();
                    }
                    element = (Element) node;
                }
                if (!htmlResult.toString().endsWith("</p>"))
                    htmlResult.append("<br>");
                return htmlResult.toString();
            }
        }
        return null;
    }

    public static boolean isXkcdImageLink(String url) {
        final String regex = "^https?://www\\.explainxkcd\\.com/wiki/index\\.php/\\d+(?!#)[:]?\\w+$";
        return url.matches(regex);
    }

    public static long getXkcdIdFromImageLink(String url) {
        final String regex = "^https?://www\\.explainxkcd\\.com/wiki/index\\.php/(\\d+).*$";
        Matcher matcher = Pattern.compile(regex).matcher(url);
        if (matcher.find()) {
            return Long.valueOf(matcher.group(1));
        } else {
            return 0;
        }
    }

    private static boolean isH2ByType(Element element, String type) {
        if (!"h2".equals(element.nodeName())) {
            return false;
        }
        for (Node child : element.childNodes()) {
            if (type.equals(child.attr("id"))) {
                return true;
            }
        }
        return false;
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/util/XkcdExplainUtil.kt
package xyz.jienan.xkcd.model.util

import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import xyz.jienan.xkcd.Const.URI_XKCD_EXPLAIN_EDIT
import java.io.IOException
import java.util.regex.Pattern

object XkcdExplainUtil {

    @Throws(IOException::class)
    fun getExplainFromHtml(responseBody: ResponseBody, url: String): String? {
        val doc = Jsoup.parse(responseBody.string())
        doc.setBaseUri(url)

        // https://explainxkcd.com/wiki/index.php/2408:_Egg_Strategies doesn't have Explanation span, good luck
        val h2Explain = doc.selectFirst("h2:has(span#Explanation)") ?: doc.selectFirst("h2")
        val nextH2Element = h2Explain.nextElementSiblings().select("h2").first()

        val h2ExplainElementIndex = h2Explain.parent().childNodes().indexOf(h2Explain)
        val nextH2ElementIndex = h2Explain.parent().childNodes().indexOf(nextH2Element)

        val explainNodes = h2Explain.parent().childNodes().subList(h2ExplainElementIndex + 1, nextH2ElementIndex)

        explainNodes.filterIsInstance<Element>().map { it.allElements }.forEach { it.cleanUp() }

        return explainNodes.joinToString("").updateEndPadding()
    }

    fun isXkcdImageLink(url: String): Boolean {
        val regexExplain = "^https?://www\\.explainxkcd\\.com/wiki/index\\.php/\\d+(?!#)[:]?\\w+$"
        val regexXkcd = "^https?://(?:www\\.|m\\.)?xkcd\\.com/\\d+/?$"
        val extraXkcd = "^https?://(?:imgs\\.)?xkcd\\.com/comics/.*$"
        return url.matches(regexExplain.toRegex()) || url.matches(regexXkcd.toRegex()) || url.matches(extraXkcd.toRegex())
    }

    fun getXkcdIdFromExplainImageLink(url: String): Long {
        val regex = "^https?://www\\.explainxkcd\\.com/wiki/index\\.php/(\\d+).*$"
        val matcher = Pattern.compile(regex).matcher(url)
        return if (matcher.find()) {
            matcher.group(1)!!.toLong()
        } else {
            0
        }
    }

    fun getXkcdIdFromXkcdImageLink(url: String): Long {
        val regex = "^https?://(?:www\\.|m\\.)?xkcd\\.com/(\\d+)/?$"
        val matcher = Pattern.compile(regex).matcher(url)
        return if (matcher.find()) {
            matcher.group(1)!!.toLong()
        } else {
            -1
        }
    }

    private fun Elements.cleanUp() {
        map { it.allElements }.forEach { elements ->
            elements.select("h3 .editsection, h3 .mw-editsection, h4 .editsection, h4 .mw-editsection").remove()
            elements.select("p sup").filter { it.toString().contains("<i>citation needed</i>") }.map { it.remove() }
            elements.select("a[href]").forEach { it.refillToFullUrl() }
            elements.select("tbody").flatMap { it.children() }.map { it.append("<br />") }
            elements.select("dd").map { it.tagName("p") }
        }
    }

    private fun Element.refillToFullUrl() {
        val href = attr("href")
        if (!href.isNullOrEmpty()) {
            if (href.startsWith("/wiki") || href.startsWith("#")) {
                attr("href", absUrl("href"))
            } else if (href.startsWith("//www.explainxkcd") && href.endsWith("action=edit")) {
                attr("href", URI_XKCD_EXPLAIN_EDIT)
            }
        }
    }

    private fun String.updateEndPadding(): String {
        return if (!endsWith("p> ")) {
            plus("<br />")
        } else {
            this
        }
    }
}